Create Patch directly from float8[]

tsl-karlp opened the pull request:
As a convenience (and for performance reasons), it's nice to be able to convert a `float8[]` directly to a Patch.

Feedback appreciated, especially the tests.

elemoine commented:
That sounds good to me. I think I'd name the function `PC_MakePatch` for consistency with the existing `PC_MakePoint` function.

elemoine commented on pgsql/pc_access.c:
```diff
@@ -329,6 +330,62 @@ Datum pcpatch_from_pcpoint_array(PG_FUNCTION_ARGS)
+	PCPOINTLIST *pl;
+	SERIALIZED_PATCH *serpa;
+	int ndims;
+	int i;
```
Indent

elemoine commented on pgsql/pc_access.c:
```diff
@@ -329,6 +330,62 @@ Datum pcpatch_from_pcpoint_array(PG_FUNCTION_ARGS)
+	int ndims;
+	int i;
+
+	PCSCHEMA *schema = pc_schema_from_pcid(pcid, fcinfo);
```
Ditto regarding indent.

elemoine commented on pgsql/pc_access.c:
```diff
@@ -329,6 +330,62 @@ Datum pcpatch_from_pcpoint_array(PG_FUNCTION_ARGS)
+        elog(ERROR, "float8[] must not have null elements");
+
+    ndims = schema->ndims;
+	nelems = ARR_DIMS(arrptr)[0];
```
Ditto.

elemoine commented on pgsql/pc_access.c:
```diff
@@ -329,6 +330,62 @@ Datum pcpatch_from_pcpoint_array(PG_FUNCTION_ARGS)
+	}
+
+    vals = (float8*) ARR_DATA_PTR(arrptr);
+	pl = pc_pointlist_make(nelems);
```
Ditto.

elemoine commented on pgsql/pc_access.c:
```diff
@@ -329,6 +330,62 @@ Datum pcpatch_from_pcpoint_array(PG_FUNCTION_ARGS)
+    vals = (float8*) ARR_DATA_PTR(arrptr);
+	pl = pc_pointlist_make(nelems);
+
+	for(i = 0; i < numpoints; ++i) {
```
Missing spaces:

```c
for ( i = 0; i < numpoints; ++i ) {
```

elemoine commented on pgsql/pointcloud.sql.in:
```diff
@@ -461,6 +461,10 @@ CREATE OR REPLACE FUNCTION PC_Patch(pcpoint[])
+CREATE OR REPLACE FUNCTION PC_Patch(pcid integer, vals float8[])
+    RETURNS pcpatch AS 'MODULE_PATHNAME', 'pcpatch_from_float_array'
+    LANGUAGE 'c' IMMUTABLE STRICT;
+
```
I'd suggest that the new SQL function definition is placed outside the "aggregate point" section.

elemoine commented on pgsql/pc_access.c:
```diff
@@ -329,6 +330,62 @@ Datum pcpatch_from_pcpoint_array(PG_FUNCTION_ARGS)
+	vals = (float8*) ARR_DATA_PTR(arrptr);
+	pl = pc_pointlist_make(nelems);
+
+	for(i = 0; i < numpoints; ++i) {
```
Please add spaces:

```
    for ( i = 0; i < numpoints; ++i ) {
```

elemoine commented on pgsql/pc_access.c:
```diff
@@ -329,6 +330,62 @@ Datum pcpatch_from_pcpoint_array(PG_FUNCTION_ARGS)
+		pc_pointlist_add_point(pl, pt);
+	}
+
+
```
(nit) please remove the extra blank line

elemoine commented on lib/pc_point.c:
```diff
@@ -217,10 +217,12 @@ pc_point_to_string(const PCPOINT *pt)
+
 PCPOINT *
-pc_point_from_double_array(const PCSCHEMA *s, double *array, uint32_t nelems)
+pc_point_from_double_array(const PCSCHEMA *s, double *array, uint32_t offset, uint32_t nelems)
```
Maybe minor but I'd rather use offset and stride here, stride being the the number of values per point.

elemoine commented:
Also, we need to add the new function to the README.

tsl-karlp commented:
@elemoine : thanks for the feedback.

* added function to the README
* fixed formatting
* renamed variable to `strike`

also, you have edit privileges to the PR, so feel to modify as you wish. I think you can squash at merge time using the GitHub interface, if you use that for merging.

elemoine commented on README.md:
```diff
@@ -251,6 +251,18 @@ Now that you have created two tables, you'll see entries for them in the `pointc
 >     SELECT PC_Patch(pt) FROM points GROUP BY id/10;
 
+
+**PC_Patch(pcid integer, vals float8[])** returns **pcpatch**
```
`PC_MakePoint`

tsl-karlp replied:
I think you meant `PC_MakePatch`, but your greater point stands. :)

elemoine commented on lib/pc_point.c:
```diff
@@ -229,9 +231,10 @@ pc_point_from_double_array(const PCSCHEMA *s, double *array, uint32_t nelems)
 
-	if ( s->ndims != nelems )
+	ndims = s->ndims;
+	if ( stride % ndims != 0 )
```
I think the test should be `if ( stride != ndims )` here. I'll make the change.

elemoine commented:
Replaced by https://github.com/pgpointcloud/pointcloud/pull/220.
